File: emoji/src/test/java/com/vanniktech/emoji/EmojiUtilsTest.java
package com.vanniktech.emoji;

import android.support.annotation.NonNull;
import com.pushtorefresh.private_constructor_checker.PrivateConstructorChecker;
import com.vanniktech.emoji.emoji.Emoji;
import com.vanniktech.emoji.emoji.EmojiCategory;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.annotation.Config;

import static org.assertj.core.api.Java6Assertions.assertThat;

@Config(manifest = Config.NONE) @RunWith(RobolectricTestRunner.class) public class EmojiUtilsTest {
  private static final String EMOJI_1 = "\u1234";
  private static final String EMOJI_2 = "\u4321";

  @Before public void setUp() {
    EmojiManager.install(new EmojiProvider() {
      @NonNull @Override public EmojiCategory[] getCategories() {
        return new EmojiCategory[]{new EmojiCategory() {
          @NonNull @Override public Emoji[] getEmojis() {
            return new Emoji[]{
                  new Emoji(EMOJI_1.codePointAt(0), R.drawable.emoji_recent),
                  new Emoji(EMOJI_2.codePointAt(0), R.drawable.emoji_backspace),
            };
          }

          @Override public int getIcon() {
            return R.drawable.emoji_recent;
          }
        }};
      }
    });
  }

  @Test public void constructorShouldBePrivate() {
    PrivateConstructorChecker.forClass(EmojiUtils.class)
          .expectedTypeOfException(AssertionError.class)
          .expectedExceptionMessage("No instances.")
          .check();
  }

  @Test public void isOnlyEmojisEmpty() {
    assertThat(EmojiUtils.isOnlyEmojis("")).isFalse();
  }

  @Test public void isOnlyEmojisNull() {
    assertThat(EmojiUtils.isOnlyEmojis(null)).isFalse();
  }

  @Test public void isOnlyEmojisText() {
    assertThat(EmojiUtils.isOnlyEmojis("hello world!")).isFalse();
  }

  @Test public void isOnlyEmojisSingleEmoji() {
    assertThat(EmojiUtils.isOnlyEmojis(EMOJI_1)).isTrue();
  }

  @Test public void isOnlyEmojisMultipleEmojis() {
    assertThat(EmojiUtils.isOnlyEmojis(EMOJI_1 + EMOJI_2)).isTrue();
  }

  @Test public void isOnlyEmojisMultipleEmojisWithSpaces() {
    assertThat(EmojiUtils.isOnlyEmojis(" \t" + EMOJI_1 + "\t    " + EMOJI_2 + "  \r\n")).isTrue();
  }

  @Test public void isOnlyEmojisSingleEmojiAndText() {
    assertThat(EmojiUtils.isOnlyEmojis(EMOJI_1 + "hello")).isFalse();
  }

  @Test public void isOnlyEmojisSingleTextAndEmoji() {
    assertThat(EmojiUtils.isOnlyEmojis("hello" + EMOJI_1)).isFalse();
  }

  @Test public void isOnlyEmojisMultipleEmojisAndText() {
    assertThat(EmojiUtils.isOnlyEmojis(EMOJI_1 + "hello" + EMOJI_2)).isFalse();
  }

  @Test public void isOnlyEmojisMultipleTextAndEmojis() {
    assertThat(EmojiUtils.isOnlyEmojis("hello" + EMOJI_1 + "world")).isFalse();
  }

  @Test public void emojisCountEmpty() {
    assertThat(EmojiUtils.emojisCount("")).isEqualTo(0);
  }

  @Test public void emojisCountNull() {
    assertThat(EmojiUtils.emojisCount(null)).isEqualTo(0);
  }

  @Test public void emojisCountText() {
    assertThat(EmojiUtils.emojisCount("hello world!")).isEqualTo(0);
  }

  @Test public void emojisCountSingleEmoji() {
    assertThat(EmojiUtils.emojisCount(EMOJI_1)).isEqualTo(1);
  }

  @Test public void emojisCountMultipleEmoji() {
    assertThat(EmojiUtils.emojisCount(EMOJI_1 + EMOJI_2)).isEqualTo(2);
  }

  @Test public void emojisCountMultipleEmojisWithSpaces() {
    assertThat(EmojiUtils.emojisCount(" " + EMOJI_1 + "    " + EMOJI_2 + "  \n")).isEqualTo(2);
  }

  @Test public void emojisCountSingleEmojiAndText() {
    assertThat(EmojiUtils.emojisCount(EMOJI_1 + "hello")).isEqualTo(1);
  }

  @Test public void emojisCountSingleTextAndEmoji() {
    assertThat(EmojiUtils.emojisCount("hello" + EMOJI_1)).isEqualTo(1);
  }

  @Test public void emojisCountMultipleEmojisAndText() {
    assertThat(EmojiUtils.emojisCount(EMOJI_1 + "hello" + EMOJI_2)).isEqualTo(2);
  }

  @Test public void emojisCountMultipleTextAndEmojis() {
    assertThat(EmojiUtils.emojisCount("hello" + EMOJI_1 + "world")).isEqualTo(1);
  }
}

File: emoji/src/commonTest/kotlin/com/vanniktech/emoji/EmojiUtilsTest.kt
/*
 * Copyright (C) 2016 - Niklas Baudy, Ruben Gees, Mario Đanić and contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.vanniktech.emoji

import kotlin.test.BeforeTest
import kotlin.test.Test
import kotlin.test.assertEquals

class EmojiUtilsTest {
  @BeforeTest fun setUp() {
    EmojiManager.install(TestEmojiProvider)
  }

  @Test fun isOnlyEmojisEmpty() {
    assertEquals(expected = false, actual = "".isOnlyEmojis())
  }

  @Test fun isOnlyEmojisNull() {
    assertEquals(expected = false, actual = null.isOnlyEmojis())
  }

  @Test fun isOnlyEmojisText() {
    assertEquals(expected = false, actual = "hello world!".isOnlyEmojis())
  }

  @Test fun isOnlyEmojisSingleEmoji() {
    assertEquals(expected = true, actual = "🎈".isOnlyEmojis())
  }

  @Test fun isOnlyEmojisMultipleEmojis() {
    assertEquals(expected = true, actual = "🎈🪀".isOnlyEmojis())
  }

  @Test fun isOnlyEmojisMultipleEmojisWithSpaces() {
    assertEquals(expected = true, actual = """ 	🎈	    🪀  """.isOnlyEmojis())
  }

  @Test fun isOnlyEmojisSingleEmojiAndText() {
    assertEquals(expected = false, actual = "🎈hello".isOnlyEmojis())
  }

  @Test fun isOnlyEmojisSingleTextAndEmoji() {
    assertEquals(expected = false, actual = "hello🎈".isOnlyEmojis())
  }

  @Test fun isOnlyEmojisMultipleEmojisAndText() {
    assertEquals(expected = false, actual = "🎈hello 🪀".isOnlyEmojis())
  }

  @Test fun isOnlyEmojisMultipleTextAndEmojis() {
    assertEquals(expected = false, actual = "hello🎈world".isOnlyEmojis())
  }

  @Test fun emojisCountEmpty() {
    assertEquals(expected = 0, actual = "".emojisCount())
  }

  @Test fun emojisCountNull() {
    assertEquals(expected = 0, actual = null.emojisCount())
  }

  @Test fun emojisCountText() {
    assertEquals(expected = 0, actual = "hello world!".emojisCount())
  }

  @Test fun emojisCountSingleEmoji() {
    assertEquals(expected = 1, actual = "🎈".emojisCount())
  }

  @Test fun emojisCountMultipleEmoji() {
    assertEquals(expected = 2, actual = "🎈🪀".emojisCount())
  }

  @Test fun emojisCountMultipleEmojisWithSpaces() {
    assertEquals(expected = 2, actual = """ 🎈    🪀  """.emojisCount())
  }

  @Test fun emojisCountSingleEmojiAndText() {
    assertEquals(expected = 1, actual = ("🎈hello").emojisCount())
  }

  @Test fun emojisCountSingleTextAndEmoji() {
    assertEquals(expected = 1, actual = ("hello🎈").emojisCount())
  }

  @Test fun emojisCountMultipleEmojisAndText() {
    assertEquals(expected = 2, actual = "🎈hello🪀".emojisCount())
  }

  @Test fun emojisCountMultipleTextAndEmojis() {
    assertEquals(expected = 1, actual = ("hello🎈world").emojisCount())
  }
}
